File: 2021/02/02.exs
input =
  File.stream!("input")
  |> Stream.map(&String.split(&1, " "))
  |> Stream.map(fn [ins, amt] -> {ins, Integer.parse(amt)} end)
  |> Stream.map(fn {ins, {amt, _}} -> {ins, amt} end)

Enum.reduce(input, {0, 0}, fn
  {"down", x}, {h, d} -> {h, d + x}
  {"up", x}, {h, d} -> {h, d - x}
  {"forward", x}, {h, d} -> {h + x, d}
end)
|> Tuple.product()
|> IO.puts()

Enum.reduce(input, {0, 0, 0}, fn
  {"down", x}, {h, d, a} -> {h, d, a + x}
  {"up", x}, {h, d, a} -> {h, d, a - x}
  {"forward", x}, {h, d, a} -> {h + x, d + (x * a), a}
end)
|> Tuple.delete_at(2)
|> Tuple.product()
|> IO.puts()

File: 2021/02/02.js
let input = require('fs').readFileSync('input', 'utf8').split('\n').filter(x => x).map(x => x.split(' ')).map(([x, y]) => [x[0], +y])

console.log(input.reduce(([h,d], [i,x]) => ({d: _ => [h,d+x], u: _ => [h,d-x], f: _ => [h+x,d]}[i]()), [0,0,0]).reduce((a,b)=>a*b))

console.log(input.reduce(([h,d,a], [i,x]) => ({d: _ => [h,d,a+x], u: _ => [h,d,a-x], f: _ => [h+x,d+(a*x),a]}[i]()), [0,0,0]).slice(0,2).reduce((a,b)=>a*b))
